filters/apiusagemonitoring/spec.go

```diff
@@ -52,15 +54,24 @@ func NewApiUsageMonitoring(
 			clientKeyList = append(clientKeyList, strippedKey)
 		}
 	}
-	// compile realms regex
-	realmsTrackingMatcher, err := regexp.Compile(realmsTrackingPattern)
+
+	var err error
+	var realmsTrackingMatcher *regexp.Regexp
+	realmsTrackingMatcherI, ok := regCache.Load(realmsTrackingPattern)
+	if !ok {
+		// compile realms regex
+		realmsTrackingMatcher, err = regexp.Compile(realmsTrackingPattern)
+	} else {
+		realmsTrackingMatcher = realmsTrackingMatcherI.(*regexp.Regexp)
+	}
```

Comment:
This strongly looks like a pattern that can be extracted into a common function `loadOrCreateRegex(pattern string, *regexp.Regexp alt) (*regexp.Regexp, Error)` for all three cases.

Reply:
makes sense!

Reply:
this load-compile-store logic appears 3 times, let's extract it to a single function

Reply:
I am on it

Reply:
pushed a fix